File: apps/server/src/routes/databases.ts
import { Elysia, t } from 'elysia'
import { requireAuth } from '$src/auth/gate'
import { getConnection } from '$src/store/connections'
import { getDriver } from '$src/db/registry'

export const databasesRoutes = new Elysia({ prefix: '/connections' })
  .use(requireAuth)
  .resolve(async ({ params, status }) => {
    const id = (params as { id?: string }).id
    if (!id || !getConnection(id)) return status(404, { error: 'Connection not found' })
    return { driver: await getDriver(id) }
  })
  .post(
    '/:id/databases',
    async ({ driver, body }) => {
      // CREATE DATABASE cannot run inside a transaction; ddl.exec uses the simple
      // protocol so this executes standalone.
      await driver.ddl.exec(`CREATE DATABASE ${driver.ddl.quoteIdent(body.name)}`)
      return { created: true as const, name: body.name }
    },
    { params: t.Object({ id: t.String() }), body: t.Object({ name: t.String({ minLength: 1 }) }) },
  )
  .delete(
    '/:id/databases/:name',
    async ({ driver, params }) => {
      await driver.ddl.exec(`DROP DATABASE ${driver.ddl.quoteIdent(params.name)}`)
      return { dropped: true as const }
    },
    { params: t.Object({ id: t.String(), name: t.String() }) },
  )

File: apps/server/src/routes/databases.rs
use std::sync::Arc;
use axum::{
    extract::{Path, State},
    routing::{delete, post},
    Json, Router,
};

pub use geto_core::services::databases::{
    CreateDatabaseRequest, CreateDatabaseResponse, DropDatabaseResponse,
};
use crate::error::AppError;
use crate::state::AppState;

#[utoipa::path(
    post,
    path = "/api/connections/{id}/databases",
    params(
        ("id" = String, Path, description = "Connection ID")
    ),
    request_body = CreateDatabaseRequest,
    responses(
        (status = 200, description = "Create new database", body = CreateDatabaseResponse),
        (status = 400, description = "Invalid database name")
    )
)]
pub async fn create_database_handler(
    State(state): State<Arc<AppState>>,
    Path(id): Path<String>,
    Json(body): Json<CreateDatabaseRequest>,
) -> Result<Json<CreateDatabaseResponse>, AppError> {
    let res = geto_core::services::databases::create_database(
        &state.registry,
        &state.core,
        &id,
        &body.name,
    )
    .await?;
    Ok(Json(res))
}

#[utoipa::path(
    delete,
    path = "/api/connections/{id}/databases/{name}",
    params(
        ("id" = String, Path, description = "Connection ID"),
        ("name" = String, Path, description = "Database name")
    ),
    responses(
        (status = 200, description = "Drop database", body = DropDatabaseResponse),
        (status = 400, description = "Invalid database name")
    )
)]
pub async fn drop_database_handler(
    State(state): State<Arc<AppState>>,
    Path((id, name)): Path<(String, String)>,
) -> Result<Json<DropDatabaseResponse>, AppError> {
    let res = geto_core::services::databases::drop_database(
        &state.registry,
        &state.core,
        &id,
        &name,
    )
    .await?;
    Ok(Json(res))
}

pub fn router() -> Router<Arc<AppState>> {
    Router::new()
        .route("/connections/{id}/databases", post(create_database_handler))
        .route("/connections/{id}/databases/{name}", delete(drop_database_handler))
}
